### PR TITLE
Automating ETL pipline Operations

### PR DESCRIPTION
**Pull request by SatwikSShanbhag**
A python script to Automate ETL pipeline operations
refer document for more understanding
https://docs.google.com/document/d/1ir_QHCac5QM17lLPhz4sW0cV3kwrWcq8K_6Zdk91eMY/edit#heading=h.uxdjawmdm40n

### CODE REVIEW

File: automate/scripts/docker-compose-hello-world.yaml
Comment (vbothe23) [inline automate/scripts/docker-compose-hello-world.yaml:1]: Why this hello-world image is requried?

File: automate/scripts/README.md [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+  1. Install and start docker.
+
+  2. Clone the repository:
+     
+     ```bash
+     git clone https://github.com/DigitalHealthIntegration/ETL-German-FHIR-Core.git
+     ```
+
+  4. Change directory to deploy. eg: cd deploy.
```
Comment (vbothe23) [inline automate/scripts/README.md]: Is 3rd point missing?

File: automate/scripts/README.md [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+             python etl.py set --vocab 5.1
+           ```
+
+  2. Set up Hapi server:
+       - Use:
+           ```bash
+             python etl.py set --hapi
+           ```
+
+  4. Download synthetic data and upload data to Hapi server
```
Comment (vbothe23) [inline automate/scripts/README.md]: Same as above

File: automate/scripts/etl_utils.py [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+        bundle_validator(bundle_data)
+        print("Bundle Validated Successfully")
+        return True
+    except ValueError as e:
+        print("Validation errors:")
+        print(e)
+        return False
+
+    except Exception as e:
+        print(f"An error occurred: {e}")
```
Comment (vbothe23) [inline automate/scripts/etl_utils.py:76]: Suggestion: If you want to print the stack trace then use `traceback.print_exc()`

File: automate/scripts/etl_utils.py [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+        return False
+
+def process_and_upload_file(file_path):
+    #fhir server endpoint
+    URL = "http://localhost:8080/fhir/"
```
Comment (vbothe23) [inline automate/scripts/etl_utils.py]: nit: Move this to configurable file like `.env`

File: automate/scripts/etl_utils.py [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+    if(os.path.isabs(file_path)):
+        full_path = file_path
+    else:
+        full_path = os.path.join(script_directory, file_path)
+    print(full_path)
+    # full_path = os.path.join(script_directory, relative_path)
+
+    for dirpath, dirnames, files in os.walk(full_path):
+
+        pattern = os.path.join(full_path, "hospitalInformation*.json")
```
Comment (vbothe23) [inline automate/scripts/etl_utils.py]: Move to string constant.

File: automate/scripts/etl_utils.py [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+#         # Download the file from S3
+#         s3.download_file(bucket_name, object_name, local_file_path)
+#         print(f"File '{object_name}' downloaded from bucket '{bucket_name}' to '{local_file_path}'")
+#         with zipfile.ZipFile(local_file_path, 'r') as zip_ref:
+#             print("extracting....")
+#             zip_ref.extractall(os.path.dirname(local_file_path))
+        
+#         print(f"File '{local_file_path}' unzipped successfully.")
+#     except Exception as e:
+#         print(f"Error downloading or unzipping file from S3: {e}")
```
Comment (vbothe23) [inline automate/scripts/etl_utils.py]: Remove if not required.

File: automate/scripts/etl_utils.py [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+            cursor.execute("SELECT * FROM cds_cdm.load_status;")
+
+            # Fetch one row from the result set
+            row = cursor.fetchone()
+
+            # Check if the load status is "started"
+            if row and row[0] == "started":
+                print("Load status is 'started'.")
+                cursor.close
+                return True
```
Comment (vbothe23) [inline automate/scripts/etl_utils.py:198]: What if the status is not `started` and it do not throw any exception?

File: automate/scripts/etl_utils.py [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+#         # Run Docker Compose command in CMD
+#         with subprocess.Popen(f'docker-compose -f "{yaml_path}" logs ohdsi-germany > .././etl_report/full_report.txt', shell=True) as process:
+#             process.wait()  # Wait for the process to complete
+#         print("Docker logs downloaded ")
+#     except subprocess.CalledProcessError as e:
+#         print("Error running Docker Compose command:", e)
+#     except KeyboardInterrupt:
+#         print("Process interrupted by the user.")
+#         # Perform any necessary cleanup here
+#         raise
```
Comment (vbothe23) [inline automate/scripts/etl_utils.py]: Remove if not required.

File: automate/scripts/etl_utils_test.py [hunk 9]
```diff
@@ -0,0 +1,553 @@
...
+
+        # Run the function
+        run_docker_compose(yaml_path)
+
+        # Check if the containers are running
+        containers = docker_client.containers.list()
+        self.assertTrue(containers)  # Assert that there are containers running
+        for container in containers:
+            print(f"container status {container.status}")
+            self.assertEqual(container.status, "exited")  # Assert that each container is running
```
Comment (vbothe23) [inline automate/scripts/etl_utils_test.py]: nit: I think printing status is not necessary in unit test. Assertion is enough.

File: automate/synthea/files/gradlew
Comment (vbothe23) [inline automate/synthea/files/gradlew:1]: Is it auto generated file?

**Review by vbothe23 — COMMENTED** [review]
Great work!! Left few comments. Please check.

File: automate/scripts/docker-compose-hello-world.yaml
Comment (SatwikSShanbhag) [reply automate/scripts/docker-compose-hello-world.yaml:1]: I am using this docker compose to test script

File: automate/scripts/etl.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+import etl_utils
+
+def reset_etl(synthea=False,hapi=False,omop=False,vocab=None,all=False):
+    if vocab:
+        print("Reset VOCAB initiated...")
+        etl_utils.remove_docker_containers("../../deploy/docker-compose-postgress.yml")
+        etl_utils.remove_docker_containers("../../deploy/docker-compose-etl.yml")
+        etl_utils.remove_docker_volume(etl_constants.omop_postgress_volume_name)
+        verify_vocab_and_download(vocab)
+        # etl_utils.run_docker_compose("../../deploy/docker-compose-postgress.yml")
```
Comment (suriyan3) [inline automate/scripts/etl.py]: If this is not required please remove it.

File: automate/scripts/etl.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+        etl_utils.remove_docker_containers("../../deploy/docker-compose-etl.yml")
+        etl_utils.remove_docker_containers(".././hapi/docker-compose-hapi.yml")
+        etl_utils.remove_docker_containers(".././synthea/docker-compose-synthea.yml")
+        etl_utils.remove_docker_volume(etl_constants.omop_postgress_volume_name)
+        etl_utils.remove_docker_volume(etl_constants.hapi_db_volume_name)
+        etl_utils.remove_folder("../../omop-vocab")
+        etl_utils.delete_files_from_given_path(".././synthea/output/fhir")
+        etl_utils.delete_files_from_given_path(".././synthea/output/metadata")
+        # etl_utils.download_vocab_from_s3()
+        #etl_utils.run_docker_compose(".././hapi/docker-compose-hapi.yml")
```
Comment (suriyan3) [inline automate/scripts/etl.py]: Remove if not required

File: automate/scripts/etl.py [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+    etl_utils.run_etl_pipeline()
+    
+
+def main():
+    # Create an ArgumentParser object
+    parser = argparse.ArgumentParser(description='ETL Automation Script')
+
+    # Add arguments
+    parser.add_argument('action', choices=['run','reset','set'], default='run', help='Action to perform')
+    # parser.add_argument('--with-hapi', action='store_true', help='Setup HAPI image')
```
Comment (suriyan3) [inline automate/scripts/etl.py]: Remove if not required

File: automate/scripts/etl_utils.py [hunk 13]
```diff
@@ -0,0 +1,440 @@
...
+        else:
+            print(f"Error: Docker volume '{volume_name}' not found.")
+    except docker.errors.APIError as e:
+        print(f"Error removing Docker volume '{volume_name}': {e}")
+    # try:
+    #     # Run Docker volume removal command in CMD
+    #     subprocess.run(f'docker volume rm "{volume_name}"', check=True)
+    #     print("Docker volume removed successfully.")
+    # except subprocess.CalledProcessError as e:
+    #     print("Error removing Docker volume:", e)
```
Comment (suriyan3) [inline automate/scripts/etl_utils.py]: Remove if not required

File: automate/scripts/etl_utils.py [hunk 14]
```diff
@@ -0,0 +1,440 @@
...
+    if validate_file_for_fhir(file_path):
+        print(f"File contains valid Fhir Resource {file_path}")
+        print("Initiating Data Upload...")
+        with open(file_path, "r", encoding="utf8") as bundle_file:         
+                data = bundle_file.read()
+                r = requests.post(url = URL, data = data.encode("utf-8"), headers = headers)
+                if r.status_code == requests.codes.ok:  # Check if the response is successful
+                    print("Response uploaded successfully.")
+                    print("Response Content:")
+                    # print(r.text)  # Print the full response content
```
Comment (suriyan3) [inline automate/scripts/etl_utils.py]: Remove if not required

File: automate/scripts/etl_utils.py [hunk 15]
```diff
@@ -0,0 +1,440 @@
...
+        hospital_files = glob.glob(pattern)
+
+        for file_path in hospital_files:
+            process_and_upload_file(file_path)
+
+        pattern = os.path.join(full_path, "practitionerInformation*.json")
+        practitioner_files = glob.glob(pattern)
+
+        for file_path in practitioner_files:
+            process_and_upload_file(file_path)
```
Comment (suriyan3) [inline automate/scripts/etl_utils.py]:
```suggestion
  def process_files_with_pattern(full_path, pattern):
      files = glob.glob(os.path.join(full_path, pattern))
      for file_path in files:
          process_and_upload_file(file_path)
          
   process_files_with_pattern(full_path, "hospitalInformation*.json")
   process_files_with_pattern(full_path, "practitionerInformation*.json")
```

File: automate/scripts/README.md (hunk 1 above)
Comment (SatwikSShanbhag) [reply automate/scripts/README.md]: No editor was changing the next numbers automatically whenever I press enter after 1st step.
Updated the numbers thanks

File: automate/scripts/etl_utils.py (hunk 6 above)
Comment (SatwikSShanbhag) [reply automate/scripts/etl_utils.py]: done

File: automate/scripts/etl_utils.py (hunk 8 above)
Comment (SatwikSShanbhag) [reply automate/scripts/etl_utils.py]: I think we can keep this code This code can be used to store container logs. if in future we want to store it any container logs 
I will uncomment and keep

File: automate/scripts/etl_utils.py [hunk 16]
```diff
@@ -0,0 +1,440 @@
...
+    else:
+        print(f"Error: Folder '{folder_path}' does not exist.")
+
+def connect_to_database():
+    try:
+        connection = psycopg2.connect(user="ohdsi_admin_user",
+                                          password="admin1",
+                                          host="localhost",
+                                          port="5412",
+                                          database="ohdsi")
```
Comment (suriyan3) [inline automate/scripts/etl_utils.py]: nit: Move this to configurable file like .env

File: automate/scripts/etl_utils.py [hunk 17]
```diff
@@ -0,0 +1,440 @@
...
+            os.makedirs(folder_path)
+            print(f"Created folder: {folder_path}")
+        
+        #download md5_hash.txt
+        print("Started downloading hash from S3")
+        object_name = f"ocl/{folder_name}/md5_hash.txt"
+        local_file_path_md5 = os.path.join(folder_path, "md5_hash.txt")
+        download_from_s3(bucket_name, object_name, local_file_path_md5, region_name)
+    except Exception as e:
+        print(f"Error downloading files from S3: {e}")
```
Comment (suriyan3) [inline automate/scripts/etl_utils.py]: This function download_hash_from_s3 and download_latest_vocab_from_s3 looks almost similar you can optimize 

def download_file_from_s3(folder_name, object_name, local_file_name):
    bucket_name = "demo-ocl-image-store"  # Replace with your S3 bucket name
    region_name = "us-east-1"
    try:
        folder_path = f"../../{folder_name}"
        if not os.path.exists(folder_path):
            os.makedirs(folder_path)
            print(f"Created folder: {folder_path}")

        # Construct local file path
        local_file_path = os.path.join(folder_path, local_file_name)

        # Download file from S3
        print(f"Started downloading {local_file_name} from S3")
        download_from_s3(bucket_name, object_name, local_file_path, region_name)
    except Exception as e:
        print(f"Error downloading files from S3: {e}")

def download_hash_from_s3(folder_name):
    object_name = f"ocl/{folder_name}/md5_hash.txt"
    download_file_from_s3(folder_name, object_name, "md5_hash.txt")

def download_latest_vocab_from_s3(folder_name):
    object_name = f"ocl/{folder_name}/omop-vocab.zip"
    download_file_from_s3(folder_name, object_name, "omop-vocab.zip")

File: automate/scripts/etl_utils.py [hunk 18]
```diff
@@ -0,0 +1,440 @@
...
+    for line in lines:
+        if line.strip().startswith(variable_to_update + '='):
+            line = f"{variable_to_update}={new_value}\n"
+        updated_lines.append(line)
+
+    # Write the updated contents back to the .env file
+    with open(env_file_path, 'w') as file:
+        file.writelines(updated_lines)
+
+    print(f"Updated {variable_to_update} to {new_value} in {env_file_path}")
```
Comment (suriyan3) [inline automate/scripts/etl_utils.py]: you can consider handling exceptions in this function  using try, except since you are doing file IO operations

**Review by suriyan3 — APPROVED** [review]
LGTM. Please address the comments.

File: automate/scripts/etl_utils.py (hunk 14 above)
Comment (SatwikSShanbhag) [reply automate/scripts/etl_utils.py]: Thanks for noticing, uncommenting this to show the response of request

File: automate/scripts/etl_utils.py (hunk 17 above)
Comment (SatwikSShanbhag) [reply automate/scripts/etl_utils.py]: I wanted it to be separate function if I want to download only or hash I can call respective methods directly.

Comment (SatwikSShanbhag) [reply automate/scripts/etl_utils.py]: after the discussion I understood  the suggestion, updating code accordingly thanks

**Review by vbothe23 — APPROVED** [review]
LGTM!

**Pull request merged**